src/deduplication/algorithm.ts

```diff
@@ -0,0 +1,44 @@
+/* eslint-disable @typescript-eslint/no-shadow */
+/* eslint-disable extra-rules/no-commented-out-code */
+/* eslint-disable no-underscore-dangle */
+import * as EL from "@elastic/elasticsearch";
+import { defaultLog } from "@pagopa/winston-ts";
+import * as TE from "fp-ts/TaskEither";
+import * as B from "fp-ts/boolean";
+import { constVoid, pipe } from "fp-ts/lib/function";
+import { IOutputDocument } from "./elasticsearch/elasticsearch";
+import { IOutputDeduplicationService } from "./elasticsearch/service";
+export const timestampDeduplication = (
+  indexName: string,
+  document: IOutputDocument
+) => (service: IOutputDeduplicationService): TE.TaskEither<Error, void> =>
+  pipe(
+    service.get(indexName, document),
+    TE.bimap(
+      resErr => resErr,
+      docRead =>
+        pipe(
+          document._timestamp > docRead.fields._timestamp,
+          B.fold(
+            () => constVoid(),
+            () =>
+              pipe(
+                service.update(indexName, document),
+                defaultLog.taskEither.info(
+                  `Document has a greater timestamp than the one in the index => updating index`
+                ),
+                TE.map(constVoid)
+              )
+          )
+        )
+    ),
+    TE.orElseW(resErr =>
+      pipe(
+        (resErr as EL.errors.ResponseError).statusCode !== 404,
```

Comment:
maybe we should return `EL.errors.ResponseError` directly in order to avoid explicit casts or even we must write a type guard instead

Reply:
the get method is the implementation of an interface using a generic type (OutputDataRead = El.estype.Result). For future implementation it could be defined as pipe of multiple types depending of the implementations.
I can't return a specific type of the elasticsearch library directly in the interface.
Do you have any other idea ?